x-pack/test/functional/services/aiops/explain_log_rate_spikes.ts

```diff
@@ -56,9 +58,93 @@ export function ExplainLogRateSpikesProvider({ getService }: FtrProviderContext)
       await testSubjects.existOrFail(`aiopsNoWindowParametersEmptyPrompt`);
     },
 
+    async assertNoResultsFoundEmptyPromptExist() {
+      await testSubjects.existOrFail(`aiopsNoResultsFoundEmptyPrompt`);
+    },
+
+    async clickDocumentCountChart() {
+      await elasticChart.waitForRenderComplete();
+      const el = await elasticChart.getCanvas();
+
+      await browser.getActions().move({ x: 0, y: 0, origin: el._webElement }).click().perform();
+    },
+
+    async clickRerunAnalysisButton(shouldRerun: boolean) {
+      await testSubjects.clickWhenNotDisabled(
+        `aiopsRerunAnalysisButton${shouldRerun ? ' shouldRerun' : ''}`
+      );
+
+      await retry.tryForTime(30 * 1000, async () => {
+        await testSubjects.existOrFail(
+          `aiopsRerunAnalysisButton${!shouldRerun ? ' shouldRerun' : ''}`
+        );
+      });
+    },
+
+    async assertAnalysisSectionExist() {
+      await retry.tryForTime(5000, async () => {
+        await testSubjects.existOrFail(`aiopsExplainLogRateSpikesAnalysis`);
+      });
+    },
+
+    async assertRerunAnalysisButton(shouldRerun: boolean) {
+      await testSubjects.existOrFail(
+        `aiopsRerunAnalysisButton${shouldRerun ? ' shouldRerun' : ''}`
+      );
+    },
+
+    async assertProgressTitle(expectedProgressTitle: string) {
+      await testSubjects.existOrFail('aiopProgressTitle');
+      const currentProgressTitle = await testSubjects.getVisibleText('aiopProgressTitle');
+      expect(currentProgressTitle).to.be(expectedProgressTitle);
+    },
+
     async navigateToIndexPatternSelection() {
       await testSubjects.click('mlMainTab explainLogRateSpikes');
       await testSubjects.existOrFail('mlPageSourceSelection');
     },
+
+    async getBrushSelectionWidth(selector: string) {
+      const brush = await testSubjects.find(selector);
+      const brushSelectionBefore = (await brush.findAllByClassName('selection'))[0];
+      const brushSelectionRectBefore = await brushSelectionBefore._webElement.getRect();
+      return brushSelectionRectBefore.width;
+    },
+
+    async getPxForTimestamp(timestamp: number) {
+      await elasticChart.waitForRenderComplete('aiopsDocumentCountChart');
+      const chartDebugData = await elasticChart.getChartDebugData('aiopsDocumentCountChart');
+
+      const dualBrushWrapper = await testSubjects.find('aiopsDualBrush');
+      const dualBrushWrapperRect = await dualBrushWrapper._webElement.getRect();
+
+      const bars = chartDebugData?.bars?.[0].bars ?? [];
+      const barsCount = bars.length;
+      const targetDeviationBarIndex = bars.findIndex((b) => b.x === timestamp);
+
+      const targetPx = Math.round(
+        (targetDeviationBarIndex / barsCount) * dualBrushWrapperRect.width
+      );
+      const intervalPx = Math.round((1 / barsCount) * dualBrushWrapperRect.width);
```

Comment:
When you're not familiar with the details, it's not obvious what `intervalPx` represents (from the method name, it's clear what the returned `targetPx` is). Could we maybe add a few comments in this method explaining `intervalPx` and maybe also some of the calculation steps, so it's easier to maintain?

Reply:
Added some comments to clarify in [f8fec5a](https://github.com/elastic/kibana/pull/138661/commits/f8fec5a432a6975f7f3638981a2d3ea088517c2b).